### PR TITLE
Document intentional and unintentional unsigned integer overflows (wraparounds) using annotations

### PR DESCRIPTION
**Pull request by practicalswift**
* ~~Add `-fsanitize=integer` Travis job~~
* Document unsigned integer overflows (wraparounds) using annotation `ALLOW_WRAPAROUND` (intentional) or `WARNING_UNINTENTIONAL_WRAPAROUND` (unintentional).

### CODE REVIEW

**Comment by ken2812221** [comment]
Concept ACK.

File: .travis.yml [hunk 1]
```diff
@@ -117,6 +117,17 @@ jobs:
...
         BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --enable-glibc-back-compat --enable-reduce-exports --with-gui=qt5 CPPFLAGS=-DDEBUG_LOCKORDER --with-sanitizers=undefined CC=clang CXX=clang++"
+# x86_64 Linux (sanitizer: integer)
+    - stage: test
+      env: >-
+        HOST=x86_64-unknown-linux-gnu
+        PACKAGES="clang python3-zmq qtbase5-dev qttools5-dev-tools libssl1.0-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libprotobuf-dev protobuf-compiler libqrencode-dev"
+        NO_DEPENDS=1
+        RUN_BENCH=false
+        UBSAN_OPTIONS="suppressions=../contrib/sanitizers-ubsan-integer.suppressions"
+        RUN_FUNCTIONAL_TESTS=false # Disabled for now, can be combined with the other x86_64 linux NO_DEPENDS job when functional tests pass the sanitizers
```
Comment (ken2812221) [inline .travis.yml]: We can't pass the functional tests with integer sanitizer even we suppress those cases?

Comment (practicalswift) [reply .travis.yml]: I got it working locally but not on Travis which is a bit weird. I'm still investigating – the intention is to run the functional tests under the sanitizer :-)

**Comment by arvidn** [comment]
there are a few `#include<>` directives that changed order, was that deliberate?
I take it the idea is to make new code that relies on unsigned wrap-around semantics to be annotated, and current annotations to be removed as code is fixed. Is that right?
would it make sense to use a different name for the macro? something like: `ALLOW_WRAPAROUND`.

nit: technically unsigned integers don't overflow, all operators on unsigned have modulo-2 arithmetic, so all results will always fit in the resulting variable.

**Comment by gmaxwell** [comment]
Please don't just blindly permit existing cases that might actually be bugs.

**Comment by practicalswift** [comment]
@arvidn `ALLOW_WRAPAROUND` is a much better name. Updated!

Yes, the include sorting was deliberate.

Yes, the idea is to make it so that new code that relies on unsigned wrap-around semantics is annotated. I've now added the following to the developer notes to clarify:

> Do not rely on unsigned wrap-around semantics unless there are good reasons for doing so (e.g. hashing). Functions that rely on on unsigned wrap-around semantics should be annotated using `ALLOW_WRAPAROUND`.
>- *Rationale*: Code relying on unsigned wrap-around semantics is legal and well-defined but error-prone.

Looks good?

**Comment by practicalswift** [comment]
@gmaxwell Yes, _intentional_ use of wraparound semantics (e.g. hashing code) is not problematic. I'll update the developer note to make that crystal clear.

**Comment by practicalswift** [comment]
@gmaxwell 

> I don't agree that it is error prone. I'm aware of no study which shows as much

"Understanding Integer Overflow in C/C++" (Dietz, Li, Regehr and Adve, 2012):

_IOC’s instrumentation is designed to be semantically transparent for programs that conform to the C or C++ language standards, except in the case where a user requests additional checking for conforming but error-prone operations, e.g., wraparound with unsigned integer types._

:-)

**Comment by practicalswift** [comment]
@arvidn @gmaxwell Thanks for reviewing. I've now updated the developer notes and split the annotations in two as suggested by @gmaxwell: `ALLOW_WRAPAROUND` for intentional wraparounds and `WARNING_UNINTENTIONAL_WRAPAROUND` for what I assume to be unintentional wraparounds.

Could you please re-review? :-)

These are the intentional wraparounds:

```
$ git grep ALLOW_WRAPAROUND -- "*.cpp" "*.h" ":(exclude)src/attributes.h"
src/bloom.cpp:ALLOW_WRAPAROUND static inline uint32_t RollingBloomHash(unsigned int nHashNum, uint32_t nTweak, const std::vector<unsigned char>& vDataToHash) {
src/crypto/chacha20.cpp:ALLOW_WRAPAROUND void ChaCha20::Output(unsigned char* c, size_t bytes)
src/crypto/ripemd160.cpp:ALLOW_WRAPAROUND void inline Round(uint32_t& a, uint32_t b, uint32_t& c, uint32_t d, uint32_t e, uint32_t f, uint32_t x, uint32_t k, int r)
src/crypto/ripemd160.cpp:ALLOW_WRAPAROUND void Transform(uint32_t* s, const unsigned char* chunk)
src/crypto/sha1.cpp:ALLOW_WRAPAROUND void inline Round(uint32_t a, uint32_t& b, uint32_t c, uint32_t d, uint32_t& e, uint32_t f, uint32_t k, uint32_t w)
src/crypto/sha1.cpp:ALLOW_WRAPAROUND void Transform(uint32_t* s, const unsigned char* chunk)
src/crypto/sha256.cpp:ALLOW_WRAPAROUND void inline Round(uint32_t a, uint32_t b, uint32_t c, uint32_t& d, uint32_t e, uint32_t f, uint32_t g, uint32_t& h, uint32_t k)
src/crypto/sha256.cpp:ALLOW_WRAPAROUND void Transform(uint32_t* s, const unsigned char* chunk, size_t blocks)
src/crypto/sha512.cpp:ALLOW_WRAPAROUND void inline Round(uint64_t a, uint64_t b, uint64_t c, uint64_t& d, uint64_t e, uint64_t f, uint64_t g, uint64_t& h, uint64_t k, uint64_t w)
src/crypto/sha512.cpp:ALLOW_WRAPAROUND void Transform(uint64_t* s, const unsigned char* chunk)
src/hash.cpp:ALLOW_WRAPAROUND unsigned int MurmurHash3(unsigned int nHashSeed, const std::vector<unsigned char>& vDataToHash)
src/hash.cpp:ALLOW_WRAPAROUND CSipHasher& CSipHasher::Write(uint64_t data)
src/hash.cpp:ALLOW_WRAPAROUND CSipHasher& CSipHasher::Write(const unsigned char* data, size_t size)
src/hash.cpp:ALLOW_WRAPAROUND uint64_t CSipHasher::Finalize() const
src/hash.cpp:ALLOW_WRAPAROUND uint64_t SipHashUint256(uint64_t k0, uint64_t k1, const uint256& val)
src/hash.cpp:ALLOW_WRAPAROUND uint64_t SipHashUint256Extra(uint64_t k0, uint64_t k1, const uint256& val, uint32_t extra)
```

These are the unintentional wraparounds:

```
$ git grep WARNING_UNINTENTIONAL_WRAPAROUND -- "*.cpp" "*.h" ":(exclude)src/attributes.h"
src/arith_uint256.h:    WARNING_UNINTENTIONAL_WRAPAROUND base_uint& operator++()
src/bench/bench.h:    WARNING_UNINTENTIONAL_WRAPAROUND inline bool KeepRunning()
src/chain.cpp:WARNING_UNINTENTIONAL_WRAPAROUND int64_t GetBlockProofEquivalentTime(const CBlockIndex& to, const CBlockIndex& from, const CBlockIndex& tip, const Consensus::Params& params)
src/chain.h:    WARNING_UNINTENTIONAL_WRAPAROUND int Height() const {
src/core_write.cpp:WARNING_UNINTENTIONAL_WRAPAROUND std::string FormatScript(const CScript& script)
src/policy/fees.cpp:WARNING_UNINTENTIONAL_WRAPAROUND double TxConfirmStats::EstimateMedianVal(int confTarget, double sufficientTxVal,
src/prevector.h:    WARNING_UNINTENTIONAL_WRAPAROUND reverse_iterator rbegin() { return reverse_iterator(item_ptr(size() - 1)); }
src/prevector.h:    WARNING_UNINTENTIONAL_WRAPAROUND const_reverse_iterator rbegin() const { return const_reverse_iterator(item_ptr(size() - 1)); }
src/script/interpreter.cpp:WARNING_UNINTENTIONAL_WRAPAROUND bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript& script, unsigned int flags, const BaseSignatureChecker& checker, SigVersion sigversion, ScriptError* serror)
src/txmempool.cpp:WARNING_UNINTENTIONAL_WRAPAROUND void CTxMemPool::UpdateAncestorsOf(bool add, txiter it, setEntries &setAncestors)
src/txmempool.cpp:WARNING_UNINTENTIONAL_WRAPAROUND void CTxMemPoolEntry::UpdateDescendantState(int64_t modifySize, CAmount modifyFee, int64_t modifyCount)
src/txmempool.cpp:WARNING_UNINTENTIONAL_WRAPAROUND void CTxMemPoolEntry::UpdateAncestorState(int64_t modifySize, CAmount modifyFee, int64_t modifyCount, int64_t modifySigOps)
src/utilstrencodings.cpp:WARNING_UNINTENTIONAL_WRAPAROUND void SplitHostPort(std::string in, int &portOut, std::string &hostOut) {
src/validation.cpp:WARNING_UNINTENTIONAL_WRAPAROUND DisconnectResult CChainState::DisconnectBlock(const CBlock& block, const CBlockIndex* pindex, CCoinsViewCache& view)
src/validation.cpp:WARNING_UNINTENTIONAL_WRAPAROUND bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CValidationState& state, const CChainParams& chainparams, CBlockIndex** ppindex, bool fRequested, const CDiskBlockPos* dbp, bool* fNewBlock)
src/validation.cpp:WARNING_UNINTENTIONAL_WRAPAROUND bool LoadMempool(void)
```

Looks correct? :-)

**Comment by practicalswift** [comment]
@ken2812221 Now running also the functional tests and the quick bench test. Please re-review :-)

File: doc/developer-notes.md [hunk 2]
```diff
@@ -439,6 +439,14 @@ General C++
 
   - *Rationale*: This avoids memory and resource leaks, and ensures exception safety
 
+- Do not rely on unsigned wrap-around semantics unless there are good reasons for doing so (e.g. hash functions).
+  Functions that intentionally rely on unsigned wrap-around semantics should be annotated using `ALLOW_WRAPAROUND`.
+  Functions that unintentionally trigger unsigned wrap-arounds should be annotated using
+  `WARNING_UNINTENTIONAL_WRAPAROUND` while awaiting being fixed.
```
Comment (arvidn) [inline doc/developer-notes.md]: would it be unrealistic to fix all these before landing this patch, so this annotation is not needed?

Comment (practicalswift) [reply doc/developer-notes.md]: @arvidn I'm not sure TBH: the review activity in #1153 which fixes these cases has been quite low during the year. My main goal with this PR is to make it less likely that we introduce new unintentional wraparounds going forward :-)

**Comment by practicalswift** [comment]
I've now changed this PR to only add annotations.

The Travis checking is added in #14252 independently of the annotations.

Also rebased.

**Comment by practicalswift** [comment]
Rebased! :-)

**Comment by practicalswift** [comment]
Rebased! :-)

**Comment by practicalswift** [comment]
Rebased!

**Comment by practicalswift** [comment]
Rebased!

**Pull request closed without merging**